src/fullfile.c

```diff
@@ -125,20 +122,88 @@ static bool download_successful(void *data)
 int download_fullfiles(struct list *files, int *num_downloads)
 {
 	void *download_handle;
+	struct list *iter;
+	struct list *need_download = NULL;
+	struct file *file;
+	struct stat stat;
+	struct download_progress download_progress = { 0, 0, 0 };
+	unsigned int complete = 0;
+	unsigned int list_length;
+	const unsigned int MAX_FILES = 1000;
 
 	if (!files) {
+		/* nothing needs to be downloaded */
 		return SWUPD_OK;
 	}
 
+	/* make a new list with only the files we actually need to download */
+	for (iter = list_head(files); iter; iter = iter->next) {
+		char *targetfile;
+		file = iter->data;
+
+		if (file->is_deleted || file->do_not_update) {
+			continue;
+		}
+
+		string_or_die(&targetfile, "%s/staged/%s", state_dir, file->hash);
+		if (lstat(targetfile, &stat) != 0 || !verify_file(file, targetfile)) {
+			need_download = list_append_data(need_download, file);
+		}
+
+		free_string(&targetfile);
+	}
+
+	if (!need_download) {
+		/* no file needs to be downloaded */
+		info("No extra files need to be downloaded\n");
+		progress_complete_step();
+		return 0;
+	}
+
+	/* getting the size of many files can be very expensive, so if
+	 * the files are not too many, get their size, otherwise just use their count
+	 * to report progress */
+	list_length = list_len(need_download);
+	if (list_length < MAX_FILES) {
+		download_progress.total_download_size = fullfile_query_total_download_size(need_download);
+		if (download_progress.total_download_size < 0) {
+			debug("Couldn't get the size of the files to download, using number of files instead\n");
+			download_progress.total_download_size = 0;
+		}
+	}
+
+	/* we need to download some files, so set up curl */
 	download_handle = swupd_curl_parallel_download_start(get_max_xfer(MAX_XFER));
 	swupd_curl_parallel_download_set_callbacks(download_handle, download_successful, download_error, NULL);
-	info("Starting download of remaining update content. This may take a while...\n");
+	swupd_curl_parallel_download_set_progress_callbacks(download_handle, swupd_progress_callback, &download_progress);
 	if (!download_handle) {
 		/* If we hit this point, the network is accessible but we were
 		 * unable to download the needed files. This is a terminal error
 		 * and we need good logging */
 		return -SWUPD_COULDNT_DOWNLOAD_FILE;
 	}
 
-	return download_loop(download_handle, files, num_downloads);
+	/* download loop */
+	info("Starting download of remaining update content. This may take a while...\n");
+
+	for (iter = list_head(need_download); iter; iter = iter->next) {
+		file = iter->data;
+
+		if (file->is_mix) {
+			download_mix_file(file);
+		} else {
+			download_file(download_handle, file);
+		}
+
+		/* fall back for progress reporting when the download size
+		* could not be determined */
+		if (download_progress.total_download_size == 0) {
+			complete++;
+			progress_report(complete, list_length);
```

Comment:
can you mix the progress_report here with progress report from the swupd_progress_callback()?

Reply:
The swupd_progress_callback() will only be used if the progress_cb was set in the swupd_curl_parallel_handle and if the total_download_size was set to something other than 0 in the data sent to the function. This function is called asynchronously by curl: "This function gets called by libcurl with a frequent interval. While data is being transferred it will be called very frequently, and during slow periods like when nothing is being transferred it can slow down to about one call per second.".

The progress_report here on fullfile.c is the fallback when swupd_progress_callback() is not being used, and in this case is used synchronously after queuing each file to be downloaded, so I don't think they can get mixed.

Reply:
ok, if the total_download_size is always zero when curl isn't reporting progress, it's fine.